frame.rkt

```diff
@@ -973,195 +976,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>."
-              (> zoomed-img-w client-w)
-              (> zoomed-img-h client-h))))
-
-    ; zooms to a specific zoom-factor (1.0 == "no zoom"),
-    ; with optional staus bar label override
-    (define/public (zoom-to factor [status-label #f])
-      (set! fit #f) ; always make sure this is cleared when setting a new zoom level
-      (define dc (send this get-dc))
-      (send dc set-scale factor factor)
-      (configure-scrollbars factor)
-      (send this refresh-now)
-      (send (status-bar-zoom) set-label
-            (cond [status-label status-label]
-                  [(not (= factor 1.0)) (format "@ ~aX" (~r factor #:precision 2))]
-                  [else ""])))
-
-    ; zooms view by a specified increment (positive or negative)
-    (define/public (zoom-by inc)
-      (define dc (send this get-dc))
-      (define-values [cur-scale-x cur-scale-y]
-        (send dc get-scale))
-      (define new-scale
-        (clamp-real (+ cur-scale-x inc) (ivl 0.1 4.0)))
-      (send this zoom-to new-scale))
-
-    ; adjusts zoom level so the entire image fits, and at least one dimension
-    ; will be the same size as the window.
-    (define/public (zoom-to-fit)
-      (let* ([client-w (send this get-width)]
-             [client-h (send this get-height)]
-             [img-w (send image-bmp-master get-width)]
-             [img-h (send image-bmp-master get-height)]
-             [new-zoom (min (/ client-w img-w)
-                            (/ client-h img-h))])
-        (send this zoom-to new-zoom "[Fit]")
-        ; must set this *after* calling zoom-to, where it is reset to false
-        (set! fit #t)))
-
-    ; only zooms out if the image is too big to fit on either dimension
-    (define/public (center-fit)
-      (let ([client-w (send this get-width)]
-            [client-h (send this get-height)]
-            [img-w (send image-bmp-master get-width)]
-            [img-h (send image-bmp-master get-height)])
-        (cond [(or (> img-w client-w)
-                   (> img-h client-h))
-               (send this zoom-to-fit)]
-              [else (send this zoom-to 1.0)])))
-
-    (define/override (on-size width height)
-      (recenter-origin width height)
-      (if fit
-          (send this zoom-to-fit)
-          (send this refresh-now)))
-
-    (define/private (recenter-origin width height)
-      (define dc (send this get-dc))
-      (send dc set-origin
-        (/ width 2)
-        (/ height 2)))
-
-    (define/public (recenter)
-      (define-values [virtual-w virtual-h] (send this get-virtual-size))
-      (recenter-origin virtual-w virtual-h))))
-
 (ivy-canvas
  (new ivy-canvas%
       [parent ivy-frame]
       [label "Ivy Image Canvas"]
       [style '(hscroll vscroll)]
       [stretchable-height #t]
+      [focus-tag-tfield focus-tag-tfield]
+      [insert-tag-tfield-comma insert-tag-tfield-comma]
+      [status-bar-position status-bar-position]
+      [status-bar-zoom status-bar-zoom]
+      [set-fullscreen set-fullscreen]
```

Comment:
What is the purpose of passing `set-fullscreen` and `toggle-fullscreen` as initialization parameters to the canvas? Wouldn't it be more appropriate to `define/public` them as member functions of the class?

Reply:
So if you look at what those functions actually do, you couldn't make them member functions, as-is. The two callbacks you mention, for example operate on the `ivy-frame` instance, which the `ivy-canvas%` class doesn't know anything about, per se.

We could call `get-parent`, but the frame may not always be the direct parent of the canvas. I'd be pretty bummed about a solution where fullscreen stopped working depending on where the canvas was in the layout hierarchy. Perhaps `get-top-level-window` would work here?

Really, though, the only reason the canvas wants to know about any of this at all is that it's the one intercepting the fullscreen accelerator keys. What would probably be a cleaner solution would be to expose more general on-key callbacks that can be registered from the outside, instead of the canvas knowing/caring about fullscreen stuff at all.

Reply:
Ah, I see. That makes sense.

Reply:
So short answer is there's a good, technical reason for it, but there's also a couple of alternatives that might be better but will require extra work. Guidance?

Reply:
I want to say that having those be member functions is the "better" approach, since `get-top-level-window` will return the ivy frame, in this case.